File: app/windows/NeuroPulse/Session/SessionProtocol.cs
// SessionProtocol.cs — hub wire format types and Ed25519 signer.
// Mirrors SessionProtocol.swift exactly. JSON key names match Swift Codable
// auto-synthesis so the hub firmware parses Windows-uploaded protocols
// identically to iOS-uploaded ones.

using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;
using NSec.Cryptography;

namespace NeuroPulse.Session;

// MARK: - Session protocol

sealed class NPSessionProtocol
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public byte SchemaVersion { get; init; } = 1;
    public required string Name { get; init; }
    public required ModalityConfig[] Modalities { get; init; }
    public required int TotalDurationSeconds { get; init; }
    // Swift JSONEncoder encodes Date as seconds since 2001-01-01 UTC (Apple reference date).
    public double CreatedAt { get; init; } = ToAppleReferenceSeconds(DateTime.UtcNow);
    public byte Mode { get; init; } = 2; // mode2Programming

    private static readonly DateTime AppleReferenceDate = new(2001, 1, 1, 0, 0, 0, DateTimeKind.Utc);
    internal static double ToAppleReferenceSeconds(DateTime utc) => (utc - AppleReferenceDate).TotalSeconds;
}

// MARK: - Modality configs

// Mirrors ModalityConfig enum in SessionProtocol.swift.
// ModalityConfigConverter serializes each subclass as {"caseName": {...}} to
// match Swift's synthesized Codable encoding of enums with associated values.
[JsonConverter(typeof(ModalityConfigConverter))]
abstract class ModalityConfig { }

// case pbmTranscranial(PBMTranscranialConfig)
sealed class PbmTranscranialConfig : ModalityConfig
{
    public required int[] Zones { get; init; }            // active zone indices 0–4
    public required double FrequencyHz { get; init; }     // 0 = CW
    public required int DutyCyclePercent { get; init; }   // ≤25 (firmware-enforced)
    public required int DurationSeconds { get; init; }
    public required double TargetDoseJoules { get; init; } // J/cm²
}

// case pbmIntranasal(PBMIntranasalConfig)
sealed class PbmIntranasalConfig : ModalityConfig
{
    public required double FrequencyHz { get; init; }
    public required int DutyCyclePercent { get; init; }
    public required int DurationSeconds { get; init; }
}

// case eegNeurofeedback(EEGConfig)
sealed class EegConfig : ModalityConfig
{
    public required string[] EnabledChannels { get; init; }
    public int SampleRateHz { get; init; } = 500;
    public required string NeurofeedbackBand { get; init; }
    public bool ClosedLoopEnabled { get; init; } = true;
}

// case bes(BESConfig)
sealed class BesConfig : ModalityConfig
{
    public required double FrequencyHz { get; init; }        // 0.5–40 Hz
    public required double AmplitudeMilliamps { get; init; } // ≤1 mA
    public required int DurationSeconds { get; init; }
    public string Waveform { get; init; } = "sinusoidal";
}

// case tdcs(TDCSConfig)
sealed class TdcsConfig : ModalityConfig
{
    public required double AmplitudeMilliamps { get; init; } // 0.1–2 mA; 40 µC/cm² limit on safety MCU
    public required int DurationSeconds { get; init; }
    public int RampSeconds { get; init; } = 30;              // hardware-enforced
    public required string[][] ElectrodePairs { get; init; }
}

// case vnsHRV(VNSHRVConfig)
sealed class VnsHrvConfig : ModalityConfig
{
    public required double FrequencyHz { get; init; }           // 1–25 Hz
    public required double AmplitudeMilliamps { get; init; }    // ≤2 mA
    [JsonPropertyName("enableHRVBiofeedback")]
    public bool EnableHrvBiofeedback { get; init; } = true;
    public double ResonanceBreathingRateDefault { get; init; } = 6.0;
    // "protocol" is a C# keyword; serialized as "protocol" to match Swift.
    [JsonPropertyName("protocol")]
    public string HrvProtocol { get; init; } = "standalone";
}

// case neuralAudio(NeuralAudioConfig)
sealed class NeuralAudioConfig : ModalityConfig
{
    public double? BinauralBeatHz { get; init; }
    public double? IsochronicToneHz { get; init; }
    public string? NoiseType { get; init; }
    public bool EegAdaptive { get; init; } = true;
    public bool UseBoneConductionForPacer { get; init; } = true;
}

// case visualStimulation(VisualStimConfig)
sealed class VisualStimConfig : ModalityConfig
{
    public required double FrequencyHz { get; init; }       // 0.5–100 Hz
    public string Mode { get; init; } = "binocular";
    [JsonPropertyName("enableModeFInvisibleNIR")]
    public bool EnableModeFInvisibleNir { get; init; } = false;
    public double EmdrCadenceHz { get; init; } = 1.0;
}

// MARK: - JSON converter for ModalityConfig

// Produces {"pbmTranscranial": {...}} format matching Swift Codable enum synthesis.
sealed class ModalityConfigConverter : JsonConverter<ModalityConfig>
{
    public override ModalityConfig Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var prop = doc.RootElement.EnumerateObject().First();
        var raw = prop.Value.GetRawText();
        return prop.Name switch
        {
            "pbmTranscranial"   => JsonSerializer.Deserialize<PbmTranscranialConfig>(raw, options)!,
            "pbmIntranasal"     => JsonSerializer.Deserialize<PbmIntranasalConfig>(raw, options)!,
            "eegNeurofeedback"  => JsonSerializer.Deserialize<EegConfig>(raw, options)!,
            "bes"               => JsonSerializer.Deserialize<BesConfig>(raw, options)!,
            "tdcs"              => JsonSerializer.Deserialize<TdcsConfig>(raw, options)!,
            "vnsHRV"            => JsonSerializer.Deserialize<VnsHrvConfig>(raw, options)!,
            "neuralAudio"       => JsonSerializer.Deserialize<NeuralAudioConfig>(raw, options)!,
            "visualStimulation" => JsonSerializer.Deserialize<VisualStimConfig>(raw, options)!,
            _ => throw new JsonException($"Unknown ModalityConfig case: {prop.Name}")
        };
    }

    public override void Write(Utf8JsonWriter writer, ModalityConfig value, JsonSerializerOptions options)
    {
        (string key, object inner) = value switch
        {
            PbmTranscranialConfig c => ("pbmTranscranial",   c),
            PbmIntranasalConfig c   => ("pbmIntranasal",     c),
            EegConfig c             => ("eegNeurofeedback",  c),
            BesConfig c             => ("bes",               c),
            TdcsConfig c            => ("tdcs",              c),
            VnsHrvConfig c          => ("vnsHRV",            c),
            NeuralAudioConfig c     => ("neuralAudio",       c),
            VisualStimConfig c      => ("visualStimulation", c),
            _ => throw new JsonException($"Unknown ModalityConfig subtype: {value.GetType().Name}")
        };
        writer.WriteStartObject();
        writer.WritePropertyName(key);
        JsonSerializer.Serialize(writer, inner, inner.GetType(), options);
        writer.WriteEndObject();
    }
}

// MARK: - Shared JSON options

static class NpJsonOptions
{
    // camelCase keys + include nulls — matches Swift JSONEncoder default output.
    public static readonly JsonSerializerOptions Hub = new(JsonSerializerDefaults.General)
    {
        WriteIndented = false,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };
}

// MARK: - Signed protocol blob

sealed class SignedProtocolBlob
{
    // Wire format: 4-byte magic "NPPR" + 4-byte LE payload length + payload + 64-byte Ed25519 sig
    private static readonly byte[] Magic = [0x4E, 0x50, 0x50, 0x52];

    public required byte[] Payload { get; init; }   // canonical JSON of NPSessionProtocol
    public required byte[] Signature { get; init; } // 64-byte Ed25519 sig
    public required string PublicKeyFingerprint { get; init; } // 8-byte hex prefix

    public byte[] ToWireFormat()
    {
        var buf = new byte[4 + 4 + Payload.Length + 64];
        Magic.CopyTo(buf.AsSpan());
        var lenBytes = BitConverter.GetBytes((uint)Payload.Length);
        if (!BitConverter.IsLittleEndian) Array.Reverse(lenBytes);
        lenBytes.CopyTo(buf.AsSpan(4));
        Payload.CopyTo(buf.AsSpan(8));
        Signature.CopyTo(buf.AsSpan(8 + Payload.Length));
        return buf;
    }
}

// MARK: - Protocol signer

// Mirrors SessionProtocolSigner.swift.
// The Ed25519 private key is generated once at first launch, protected with
// DPAPI (Windows user-scope), and persisted to %LOCALAPPDATA%\NeuroPulse\.
// The paired hub stores the corresponding public key registered at BLE pairing.
static class SessionProtocolSigner
{
    private static readonly SignatureAlgorithm Ed25519Alg = SignatureAlgorithm.Ed25519;
    private static readonly string KeyPath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "NeuroPulse", "signing.key");

    public static SignedProtocolBlob Sign(NPSessionProtocol proto)
    {
        var payload = JsonSerializer.SerializeToUtf8Bytes(proto, NpJsonOptions.Hub);
        using var key = LoadOrCreateSigningKey();
        var sig = Ed25519Alg.Sign(key, payload);
        var fp = Convert.ToHexString(key.PublicKey.Export(KeyBlobFormat.RawPublicKey).AsSpan()[..8]).ToLowerInvariant();
        return new SignedProtocolBlob { Payload = payload, Signature = sig, PublicKeyFingerprint = fp };
    }

    public static byte[] PublicKeyData()
    {
        using var key = LoadOrCreateSigningKey();
        return key.PublicKey.Export(KeyBlobFormat.RawPublicKey);
    }

    private static Key LoadOrCreateSigningKey()
    {
        var creationParams = new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport };

        if (File.Exists(KeyPath))
        {
            var dpapi = File.ReadAllBytes(KeyPath);
            var raw = ProtectedData.Unprotect(dpapi, null, DataProtectionScope.CurrentUser);
            return Key.Import(Ed25519Alg, raw, KeyBlobFormat.RawPrivateKey, creationParams);
        }

        Directory.CreateDirectory(Path.GetDirectoryName(KeyPath)!);
        var key = Key.Create(Ed25519Alg, creationParams);
        var exported = key.Export(KeyBlobFormat.RawPrivateKey);
        var protected_ = ProtectedData.Protect(exported, null, DataProtectionScope.CurrentUser);
        File.WriteAllBytes(KeyPath, protected_);
        return key;
    }
}

File: app/ios/NeurOne/Session/SessionProtocol.swift
import Foundation
import CryptoKit

// Session protocol authoring and CSPRNG signing.
// A session protocol defines all stimulation parameters for a single session.
// It is cryptographically signed before upload to the hub; the hub firmware
// rejects unsigned or corrupted protocols (NP-FW-EMMC-001 Rev A §8, CLAUDE.md §4.2).

// MARK: - Protocol definition

struct NPSessionProtocol: Codable {
    var id: UUID = UUID()
    /// v2: cranial targeting is a socket bitmap, not five zone-slot indices
    /// (NP-HEX-ZM-001). The blob SHAPE changed, so the version had to move with
    /// it — a hub reading a v2 blob as v1 would read a 16-byte mask where it
    /// expects a zone list.
    var schemaVersion: UInt8 = 2
    var name: String
    var modalities: [ModalityConfig]
    var totalDurationSeconds: Int
    var createdAt: Date = Date()
    var mode: OperatingMode = .mode2Programming

    enum OperatingMode: UInt8, Codable {
        case mode1Connected   = 1  // real-time streaming
        case mode2Programming = 2  // pre-upload, run from hub
        case mode3Autonomous  = 3  // offline, any USB-C PD power bank
    }
}

// MARK: - Modality configurations

enum ModalityConfig: Codable {
    case pbmTranscranial(PBMTranscranialConfig)
    case pbmIntranasal(PBMIntranasalConfig)
    case eegNeurofeedback(EEGConfig)
    case bes(BESConfig)
    case tdcs(TDCSConfig)
    case vnsHRV(VNSHRVConfig)
    case neuralAudio(NeuralAudioConfig)
    case visualStimulation(VisualStimConfig)
}

struct PBMTranscranialConfig: Codable {
    /// Which sockets on the hex lattice this command drives — the firmware's
    /// NP_PROTO_TARGET_SOCKET_MASK representation (16 bytes, LSB-first, 0-based).
    /// Replaces the five zone-slot indices, which named hardware that no longer
    /// exists.
    var socketMask: NPSocketMask
    var frequencyHz: Double     // 0 = CW, >0 = pulsed
    var dutyCyclePercent: Int   // ≤25 (firmware-enforced for pulsed)
    var durationSeconds: Int
    var targetDoseJoules: Double  // J/cm²; hub closed-loop to this target
}

struct PBMIntranasalConfig: Codable {
    var frequencyHz: Double
    var dutyCyclePercent: Int
    var durationSeconds: Int
}

struct EEGConfig: Codable {
    var enabledChannels: [String]  // e.g. ["Fp1","Fp2","F3","F4","C3","C4","P3","P4"]
    var sampleRateHz: Int = 500
    var neurofeedbackBand: String  // e.g. "alpha", "theta", "gamma"
    var closedLoopEnabled: Bool = true
}

struct BESConfig: Codable {
    var frequencyHz: Double        // 0.5–40 Hz
    var amplitudeMilliamps: Double // ≤1 mA
    var durationSeconds: Int
    var waveform: String = "sinusoidal"
}

struct TDCSConfig: Codable {
    var amplitudeMilliamps: Double // 0.1–2 mA; 40 µC/cm² hard limit enforced by safety MCU
    var durationSeconds: Int
    var rampSeconds: Int = 30      // hardware-enforced in firmware
    var electrodePairs: [[String]] // e.g. [["Fp1","P3"]]
}

struct VNSHRVConfig: Codable {
    var frequencyHz: Double    // 1–25 Hz
    var amplitudeMilliamps: Double  // ≤2 mA
    var enableHRVBiofeedback: Bool = true
    var resonanceBreathingRateDefault: Double = 6.0  // breaths/min
    var hrvProtocol: HRVProtocol = .standalone

    enum HRVProtocol: String, Codable {
        case standalone            = "standalone"
        case tavnsSynchronized     = "hrv_tavns_sync"
        case dualEEGBiofeedback    = "hrv_eeg_biofeedback"
        case combinedPBM           = "hrv_pbm"
    }
}

struct NeuralAudioConfig: Codable {
    var binauralBeatHz: Double?
    var isochronicToneHz: Double?
    var noiseType: String?         // "pink", "brown", or nil
    var eegAdaptive: Bool = true
    var useBoneConductionForPacer: Bool = true
}

struct VisualStimConfig: Codable {
    var frequencyHz: Double        // 0.5–100 Hz
    var mode: String = "binocular" // "binocular", "emdr", "retinalPBM"
    var enableModeFInvisibleNIR: Bool = false
    var emdrCadenceHz: Double = 1.0
}

// MARK: - Signed protocol blob

struct SignedProtocolBlob {
    let payload: Data   // canonical JSON of NPSessionProtocol
    let signature: Data // Ed25519 signature over SHA-256(payload) + schemaVersion + payloadSize
    let publicKeyFingerprint: String  // hex, for hub key pinning

    // Wire format: 4-byte magic + 4-byte payload length + payload + 64-byte Ed25519 sig
    static let magic: [UInt8] = [0x4E, 0x50, 0x50, 0x52]  // "NPPR"

    var wireFormat: Data {
        var buf = Data(Self.magic)
        var len = UInt32(payload.count).littleEndian
        buf.append(Data(bytes: &len, count: 4))
        buf.append(payload)
        buf.append(signature)
        return buf
    }
}

// MARK: - Protocol signer

struct SessionProtocolSigner {
    // The private signing key is generated once at first launch and stored in the iOS Keychain.
    // The corresponding public key is registered with the hub at pairing time.
    // Production: replace with secure enclave key where available.

    private static let keychainTag = "life.neurone.session-signing-key"

    static func sign(_ proto: NPSessionProtocol) throws -> SignedProtocolBlob {
        let payload = try JSONEncoder().encode(proto)
        let signingKey = try loadOrCreateSigningKey()
        let digest = SHA256.hash(data: payload)
        let signature = try signingKey.signature(for: Data(digest))
        let fingerprint = Data(signingKey.publicKey.rawRepresentation).prefix(8)
            .map { String(format: "%02x", $0) }.joined()
        return SignedProtocolBlob(
            payload: payload,
            signature: Data(signature),
            publicKeyFingerprint: fingerprint
        )
    }

    private static func loadOrCreateSigningKey() throws -> Curve25519.Signing.PrivateKey {
        // Attempt to load existing key from Keychain.
        // kSecAttrAccessible must be present in the read query so a key written by an
        // attacker with a different accessibility class cannot shadow the legitimate key
        // via SecItemCopyMatching. (NP-PRIV-ANALYSIS-002 LOW-12)
        let query: [CFString: Any] = [
            kSecClass:              kSecClassKey,
            kSecAttrApplicationTag: keychainTag.data(using: .utf8)!,
            kSecAttrAccessible:     kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            kSecReturnData:         true
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        if status == errSecSuccess, let keyData = item as? Data {
            return try Curve25519.Signing.PrivateKey(rawRepresentation: keyData)
        }

        // Generate new key and store in Keychain
        let key = Curve25519.Signing.PrivateKey()
        let addQuery: [CFString: Any] = [
            kSecClass:                kSecClassKey,
            kSecAttrApplicationTag:   keychainTag.data(using: .utf8)!,
            kSecValueData:            key.rawRepresentation,
            kSecAttrAccessible:       kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        SecItemAdd(addQuery as CFDictionary, nil)
        return key
    }

    static func publicKeyData() throws -> Data {
        let key = try loadOrCreateSigningKey()
        return key.publicKey.rawRepresentation
    }
}
